File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = 0
        n = len(s)
        ans = 0 
        seen = defaultdict(int)
        
        for right in range(n):
            seen[s[right]] += 1
            while len(seen) < right - left + 1:
                seen[s[left]] -= 1
                if seen[s[left]] == 0:
                    seen.pop(s[left])
                left += 1
            
            ans = max(ans, len(seen))
        
        return ans

File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0;
        int longest = 0;
        Set<Character> seen = new HashSet<>();
        
        for(int right = 0; right < s.length(); right++){
            while(seen.contains(s.charAt(right))){
                seen.remove(s.charAt(left));
                left += 1;
            }
            seen.add(s.charAt(right));
            longest = Math.max(longest, right - left + 1);
        }
        
        return longest;
    }
}